Plutarch/Rational.hs

```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
   pforgetData,
   pfromDataImpl,
  )
-import Plutarch.Integer (PInteger, PIntegral (pdiv, pmod))
-import Plutarch.Internal (PType)
+import Plutarch.Integer (PInteger, pdiv, pmod)
 import Plutarch.Lift (pconstant)
-import Plutarch.List (PListLike (pcons, phead, pnil, ptail), pmap)
+import Plutarch.List (pcons, phead, plength, pnil, ptail)
 import Plutarch.Num (PNum, pabs, pfromInteger, pnegate, psignum, (#*), (#+), (#-))
-import Plutarch.Pair (PPair (..))
-import Plutarch.Show (PShow (pshow'), pshow)
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Positive (PPositive, ptryPositive)
+import Plutarch.Show (PShow, pshow, pshow')
+import Plutarch.TermCont (tcont, unTermCont)
 import Plutarch.Trace (ptraceError)
-import Plutarch.Unsafe (punsafeCoerce)
+import Plutarch.TryFrom (PTryFrom (PTryFromExcess, ptryFrom'), ptryFrom)
+import Plutarch.Unsafe (punsafeCoerce, punsafeDowncast)
 
 class PFractional (a :: PType) where
   (#/) :: Term s a -> Term s a -> Term s a
   precip :: Term s (a :--> a)
   pfromRational :: Term s (PRational :--> a)
 
 data PRational s
-  = PRational (Term s PInteger) (Term s PInteger)
+  = PRational (Term s PInteger) (Term s PPositive)
   deriving stock (Generic)
-  deriving anyclass (PlutusType)
+  deriving anyclass (PlutusType, PEq)
+
 instance DerivePlutusType PRational where type DPTStrat _ = PlutusTypeScott
 
 instance (PNum a, PFractional a) => Fractional (Term s a) where
   (/) = (#/)
   recip x = precip # x
-  fromRational x = pfromRational #$ pcon $ PRational (pconstant (numerator x)) (pconstant (denominator x))
+  fromRational x =
+    pfromRational #$ pcon $
+      PRational
+        (pconstant $ numerator x)
+        (punsafeDowncast . pconstant $ denominator x)
```

Comment:
Are we sure `denominator` is always positive?

Reply:
[it should be](https://hackage.haskell.org/package/base-4.16.2.0/docs/src/GHC.Real.html#%25) - but we can have further haskell level logic here to invert the sign of the numerator if denominator is negative. Probably unnecessary though.